src/Lookup.hs

```diff
@@ -30,90 +28,56 @@ lookupInEnv path@(SymPath (p : ps) name) env =
-      let envs = (if allowLookupInAllModules then importsAll else importsNormal) env
-       in mapMaybe (lookupInLocalEnv name) envs
-    recursiveLookup :: Env -> [(Env, Binder)]
-    recursiveLookup env =
-      let spine = case Map.lookup name (envBindings env) of
-            Just found -> [(env, found)]
-            Nothing -> []
-          leaves = importsLookup env
-          above = case envParent env of
-            Just parent -> recursiveLookup parent
-            Nothing -> []
-       in --(trace $ "multiLookupInternal '" ++ name ++ "' " ++ show (envModuleName env) ++ ", spine: " ++ show (fmap snd spine) ++ ", leaves: " ++ show (fmap snd leaves) ++ ", above: " ++ show (fmap snd above))
-          spine ++ leaves ++ above
+-- | Like 'lookupBinder' but return the Meta for the binder, or a default empty meta.
+lookupMeta :: SymPath -> Env -> MetaData
+lookupMeta path globalEnv =
+  case lookupBinder path globalEnv of
+    Just (Binder meta _) -> meta
+    Nothing -> emptyMeta
 
-binderToEnv :: Binder -> Maybe Env
-binderToEnv (Binder _ (XObj (Mod e) _ _)) = Just e
-binderToEnv _ = Nothing
+-- | Get the Env stored in a binder, if any.
+envFromBinder :: Binder -> Maybe Env
+envFromBinder (Binder _ (XObj (Mod e) _ _)) = Just e
+envFromBinder _ = Nothing
 
 -- | Given an environment, returns the list of all environments of binders from
--- imported modules `(load "module-file.carp")`
+-- imported modules.
 importedEnvs :: Env -> [Env]
 importedEnvs env =
-  let envs = mapMaybe (binderToEnv . snd) (Map.toList (envBindings env))
-   in envs ++ concatMap importedEnvs envs
+  catMaybes $ mapMaybe (\path -> fmap envFromBinder (lookupBinder path env)) (envUseModules env)
+
+-- | Given an environment, returns the list of all environments of its binders.
+allEnvs :: Env -> [Env]
+allEnvs env =
+  let envs = mapMaybe (envFromBinder . snd) (Map.toList (envBindings env))
+   in envs ++ concatMap allEnvs envs
+
+data LookWhere = Everywhere | OnlyImports
+
+getEnvs :: LookWhere -> Env -> [Env]
+getEnvs Everywhere = allEnvs
+getEnvs OnlyImports = importedEnvs
 
 -- | Given an environment, use a lookup function to recursively find all binders
 -- in the environment that satisfy the lookup.
-recursiveLookupAll :: a -> LookupFunc a -> Env -> [Binder]
-recursiveLookupAll input lookf env =
+lookupMany :: LookWhere -> LookupFunc a b -> a -> Env -> [b]
+lookupMany lookWhere lookf input env =
   let spine = lookf input env
-      leaves = concatMap (lookf input) (importedEnvs env)
+      leaves = concatMap (lookf input) (getEnvs lookWhere env)
       above = case envParent env of
-        Just parent -> recursiveLookupAll input lookf parent
+        Just parent -> lookupMany lookWhere lookf input parent
         Nothing -> []
    in spine ++ leaves ++ above
 
--- | Lookup binders by name.
-lookupByName :: String -> Env -> [Binder]
+-- | Lookup binders by name in a single Env (no recursion),
+lookupByName :: String -> Env -> [(Env, Binder)]
```

Comment:
If this only performs a lookup in a single env without recursing, does it need to return the `env`? After all, `env` in this case is precisely the `Env` argument passed to the function, so it doesn't provide any new information to callers.

Reply:
I think it has to, due to the generality of `lookupMany` (specifically, the signature of `lookupByName` makes it possible to construct the `multiLookup...` functions that all need to return the env where the binding was found)

Reply:
Ahhh, gotcha, thanks for pointing that out!

Reply:
I might be missing something also :)